crates/sui-graphql-rpc/src/server/builder.rs

```diff
@@ -471,10 +461,13 @@ async fn graphql_handler(
     // Note: if a load balancer is used it must be configured to forward the client IP address
     req.data.insert(addr);
 
-    let checkpoint_viewed_at = watermark.0 .0.load(Relaxed);
+    let watermark = service_watermark.read().await;
 
     // This wrapping is done to delineate the watermark from potentially other u64 types.
-    req.data.insert(CheckpointViewedAt(checkpoint_viewed_at));
+    req.data.insert(Watermark {
+        checkpoint: watermark.checkpoint,
+        epoch: watermark.epoch,
+    });
```

Comment:
Minimise the time you hold the read lock for as well:

```rust
req.data.insert({
    let w = service_watermark.read().await;
    Watermark {
      checkpoint: watermark.checkpoint,
      epoch: watermark.epoch
    }
});
```

(Or better yet, create a function on the type protected by the mutex to create a `Watermark`).